testing/qemu_riscv.md

```diff
@@ -0,0 +1,107 @@
+Setting up RISC-V emulation for testing bitcoin
+===============================================
+
+This document describes some steps to set up qemu in user mode to emulate the
+RISC-V architecture in user-space and pass through the system calls to the native
+system.
+
+This is efficient for testing because it makes it unnecessary to install a whole VM.
+
+It also makes it possible to control extactly what libraries are available to prevent
+external dependencies from leaking in.
+
+These same steps work for other architectures but for sake of simplicity, I'll
+only focus on RISC-V here. This comes up a lot for RISC-V specifically because
+it's still so hard to find hardware for it.
+
+Installing qemu
+----------------
+
+Many distributions don't provide packages for a qemu that supports RISC-V 
+as upstream support for this has been merged somewhat recently. Here are some steps
+to build one from source.
+
+Dependencies for debian:
+
+```bash
+apt install libglib2.0-dev
+```
+
+Check out the repository and build, install:
+
+```bash
+git clone https://github.com/qemu/qemu.git 
+cd qemu
+./configure --prefix=/opt/qemu --target-list=riscv64-linux-user --disable-tools
+make -j4
+make install
+```
+
+(optional) add install path to `.profile` or such:
+```bash
+echo 'PATH="/opt/qemu/bin:$PATH"' >> ~/.profile
+```
+
+RISC-V dynamic linker
+---------------------
+
+qemu will look in `/usr/gnemul/qemu-riscv64/lib` by default for the dynamic
+linker (ELF interpreter). This can be overridden with the environment variable
+`QEMU_LD_PREFIX`.
+
+    ld-2.29.9000.so -> ../lib64/ld-2.29.9000.so
+
+You can copy these files from an existing RISC-V install. Alternatively, you
+can download the dynamic linker and bare-bones shared libraries (copied from
+Fedora rawhide) here:
+
+```bash
+mkdir -p /usr/gnemul
+cd /usr/gnemul
+wget https://dev.visucore.com/bitcoin/qemu-riscv64-libs.tar.gz
+tar -zxvf qemu-riscv64-libs.tar.gz
+```
+
+(TODO: get this files directly from the packages instead of from wumpus' server)
```

Comment:
Some distributions might even already have a package that provides these, for cross-architecture. e.g. https://wiki.debian.org/RISC-V mentions:
```
$ sudo dpkg --add-architecture riscv64
$ sudo apt-get install gcc-riscv64-linux-gnu g++-riscv64-linux-gnu
$ sudo sh -c "cat >/etc/ld.so.conf.d/riscv64-linux-gnu.conf <<EOF
/usr/local/lib/riscv64-linux-gnu
/lib/riscv64-linux-gnu
/usr/lib/riscv64-linux-gnu
/usr/riscv64-linux-gnu/lib/
EOF
"
$ sudo ln -s /usr/riscv64-linux-gnu/lib/ld-linux-riscv64-lp64d.so.1 /lib
```
I haven't tried this.